gateway-sp-comms/src/single_sp.rs

```diff
@@ -0,0 +1,749 @@
+use gateway_messages::SpMessageKind;
+use gateway_messages::SpPort;
+use gateway_messages::SpState;
+use gateway_messages::UpdateChunk;
+use gateway_messages::UpdateStart;
+use omicron_common::backoff;
+use omicron_common::backoff::Backoff;
+use slog::debug;
+use slog::error;
+use slog::info;
+use slog::trace;
+use slog::warn;
+use slog::Logger;
+use std::convert::TryInto;
+use std::io;
+use std::net::Ipv6Addr;
+use std::net::SocketAddrV6;
+use std::time::Duration;
+use thiserror::Error;
+use tokio::net::UdpSocket;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use tokio::sync::watch;
+use tokio::task::JoinHandle;
+use tokio::time;
+use tokio::time::timeout;
+
+use crate::communicator::ResponseKindExt;
+use crate::error::BadResponseType;
+
+pub const DISCOVERY_MULTICAST_ADDR: Ipv6Addr =
+    Ipv6Addr::new(0xff15, 0, 0, 0, 0, 0, 0x1de, 0);
+
+// Once we've discovered an SP, continue to send discovery packets on this
+// interval to detect changes.
+//
+// TODO-correctness/TODO-security What do we do if the SP address changes?
+const DISCOVERY_INTERVAL_IDLE: Duration = Duration::from_secs(60);
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("failed to send UDP packet to {addr}: {err}")]
+    UdpSendTo { addr: SocketAddrV6, err: io::Error },
+    #[error("failed to recv UDP packet: {0}")]
+    UdpRecv(io::Error),
+    #[error("failed to deserialize SP message from {peer}: {err}")]
+    Deserialize { peer: SocketAddrV6, err: gateway_messages::HubpackError },
+    #[error("RPC call failed (gave up after {0} attempts)")]
+    ExhaustedNumAttempts(usize),
+    #[error("serial console already attached")]
+    SerialConsoleAlreadyAttached,
+    #[error(transparent)]
+    BadResponseType(#[from] BadResponseType),
+    #[error("Error response from SP: {0}")]
+    SpError(#[from] ResponseError),
+}
+
+#[derive(Debug, Error)]
+pub enum UpdateError {
+    #[error("update image is too large")]
+    ImageTooLarge,
+    #[error("error starting update: {0}")]
+    Start(Error),
+    #[error("error updating chunk at offset {offset}: {err}")]
+    Chunk { offset: u32, err: Error },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug)]
+pub struct SingleSp {
+    inner_commands: mpsc::Sender<InnerCommand>,
+    sp_addr: watch::Receiver<Option<(SocketAddrV6, SpPort)>>,
+    inner_task: JoinHandle<()>,
+    log: Logger,
+}
+
+impl Drop for SingleSp {
+    fn drop(&mut self) {
+        self.inner_task.abort();
+    }
+}
+
+impl SingleSp {
+    /// Construct a new `SingleSp` that will periodically attempt to discover an
+    /// SP reachable at `discovery_addr` (typically
+    /// [`DISCOVERY_MULTICAST_ADDR`], but possibly different in test /
+    /// development setups).
+    pub fn new(
+        socket: UdpSocket,
+        discovery_addr: SocketAddrV6,
+        max_attempts: usize,
+        per_attempt_timeout: Duration,
+        log: Logger,
+    ) -> Self {
+        let (cmd_tx, cmd_rx) = mpsc::channel(8);
```

Comment:
Totally arbitrary; the SP doesn't support pipelined requests, but this would allow a small amount of requests to be queued up. Maybe this should be 2 instead, allowing 1 enqueued request beyond the currently-in-flight one?

Reply:
Gotcha, maybe just add a comment.

Reply:
Added in 94e10401